### PR TITLE
[Refinement] common: add ensure_otaproxy_start for probing otaproxy online

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

This PR introduces `ensure_otaproxy_start` function that can be used for probing otaproxy until it is online or exceed \<timeout>.

**NOTE**: this PR is split from #212 , and intended to be used in #212, this PR only implements the function but not yet uses.

### Motivation

There is an interval between subECU starts updating and mainECU's otaproxy become online, as otaproxy has to take some time to launch and become ready, especially when there are left over cache files needed to be scrubbed. 

To reduce unnecessary downloading failures and retrying when otaproxy is not yet online, this PR introduces the `ensure_otaproxy_start` function,  which allow otaclient to probe otaproxy online before it starts any downloading.

## Check list

- [x] test file(s) that covers the change(s) is implemented.
- [x] local test is passed. 

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] No (function is only implemented, not yet in used).

## Breaking change

Does this PR introduce breaking change?

- [x] No (function is only implemented, not yet in used).

### CODE REVIEW

File: otaclient/app/common.py [hunk 1]
```diff
@@ -597,3 +598,45 @@ def map(
...
+    with requests.Session() as session:
+        while start_time + probing_timeout > (cur_time := int(time.time())):
+            try:
+                resp = session.get(otaproxy_url, timeout=connection_timeout)
+                resp.close()
+                return
+            except Exception as e:  # server is not up yet
+                if cur_time >= next_warning:
+                    logger.warning(
+                        f"otaproxy@{otaproxy_url} is not up after {warning_count * warning_interval} seconds"
```
Comment (obi-t4) [inline otaclient/app/common.py:632]: To introduce this fix, waiting_count can be removed.
```suggestion
                        f"otaproxy@{otaproxy_url} is not up after {cur_time - start_time} seconds"
```

File: otaclient/app/common.py [hunk 2]
```diff
@@ -597,3 +598,45 @@ def map(
...
+):
+    """Loop probing <otaproxy_url> until online or exceed <probing_timeout>.
+
+    This function will issue a logging.warning every <warning_interval> seconds.
+
+    Raises:
+        A ConnectionError if exceeds <probing_timeout>.
+    """
+    start_time = int(time.time())
+    warning_count, next_warning = 0, start_time + warning_interval
```
Comment (obi-t4) [inline otaclient/app/common.py:619]: and then `warning_count` should be removed.

**Comment by obi-t4** [comment]
@Bodong-Yang 
Can you follow the PR template?

**Comment by Bodong-Yang** [comment]
Sure! Also I am creating test for this PR, please wait a moment.

**Comment by obi-t4** [comment]
> Sure! Also I am creating test for this PR, please wait a moment.

I'll change this PR to draft for now.

**Comment by Bodong-Yang** [comment]
@obi-t4 The test files for this PR is implemented, please have a check.

File: tests/test_common.py [hunk 3]
```diff
@@ -380,3 +383,62 @@ def test_succeeded_in_one_try(self):
...
+            yield
+        finally:
+            _server_p.kill()
+
+    def test_timeout_waiting(self):
+        """
+        NOTE: we intentionally not use the subprocess_launch_server fixture here
+              to let the probing timeout.
+        """
+        probing_timeout = 3
```
Comment (obi-t4) [inline tests/test_common.py:422]: to understand the relationship between LAUNCH_DELAY and probing_timeout.
```suggestion
        probing_timeout = self.LAUNCH_DELAY // 2
```

File: tests/test_common.py [hunk 4]
```diff
@@ -380,3 +383,62 @@ def test_succeeded_in_one_try(self):
...
+                self.DUMMY_SERVER_URL,
+                interval=0.1,
+                connection_timeout=0.1,
+                probing_timeout=probing_timeout,
+            )
+        assert int(time.time()) >= start_time + probing_timeout
+
+    def test_probing_delayed_online_server(self, subprocess_launch_server):
+        start_time = int(time.time())
+        probing_timeout = self.LAUNCH_DELAY + 6
```
Comment (obi-t4) [inline tests/test_common.py:436]: same reason.
```suggestion
        probing_timeout = self.LAUNCH_DELAY * 2
```

**Review by obi-t4 — COMMENTED** [review]
I left some comment. Please have a check.

File: tests/test_common.py (hunk 3 above)
Comment (Bodong-Yang) [reply tests/test_common.py]: Updated in [9da9ac1](https://github.com/tier4/ota-client/pull/219/commits/9da9ac1416d1de484a4356cbd922b4c917b32dbe)

File: tests/test_common.py (hunk 4 above)
Comment (Bodong-Yang) [reply tests/test_common.py]: Updated in [9da9ac1](https://github.com/tier4/ota-client/pull/219/commits/9da9ac1416d1de484a4356cbd922b4c917b32dbe)

File: tests/test_common.py [hunk 5]
```diff
@@ -419,26 +423,28 @@ def test_timeout_waiting(self):
         NOTE: we intentionally not use the subprocess_launch_server fixture here
               to let the probing timeout.
         """
-        probing_timeout = 3
+        probing_timeout = self.LAUNCH_DELAY * 2
```
Comment (obi-t4) [inline tests/test_common.py:426]: :eye:

Comment (Bodong-Yang) [reply tests/test_common.py:426]: Ah sorry it should be // here.

Comment (Bodong-Yang) [reply tests/test_common.py]: Updated in [4f0fb0e](https://github.com/tier4/ota-client/pull/219/commits/4f0fb0ecd5021dfaaf70b7c68d6d924f92a76d3f).

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

**Pull request merged**